Adds searchbar and category tags 2.00b Electronics Library Dev

mmtywa opened the pull request:
- uses fast-fuzzy npm package for text search

mmtywa committed:
Adds searchbar and category tags 2p00b Electronics Library Dev

- uses fast-fuzzy npm package for text search

mmtywa committed:
Adds newlines at EOF

dillzhang commented on pages/electronics-library-dev.tsx:
```diff
@@ -18,7 +21,23 @@ interface IElectronicsComponent {
+};
+
+const categoryObject:{[key in CategoryTags]: {result: string, color: string}} = {
+  ShowAll: {result: "All Components", color: '#2D2D2D'},
```
Is there a reason this one is a hex code while the rest are numeric rgb values? Can we keep it consistent?

mmtywa replied:
I thought I had a reason at the time, but now I just think I'm crazy... I fixed to hex values.

dillzhang commented on pages/electronics-library-dev.tsx:
```diff
@@ -1,10 +1,13 @@
 import styles from '../styles/Page.module.scss';
 import cx from 'classnames';
 
+const {Searcher} = require("fast-fuzzy");
```
Can you double check your linter? `Searcher` should have spaces around it similar to the imports above.

mmtywa replied:
Okay I made sure linting is working on my editor. Format should be good in next commit

dillzhang commented on pages/electronics-library-dev.tsx:
```diff
@@ -18,7 +21,23 @@ interface IElectronicsComponent {
-}
+};
+
+enum CategoryTags {
```
Love the use of an enum here!

dillzhang commented on pages/electronics-library-dev.tsx:
```diff
@@ -109,6 +86,59 @@ const Home = ({
+    setSearchField(e.currentTarget.value);
+  };
+  const onTagSelect = (e: React.FormEvent<HTMLInputElement>) => {
+    setSearchCategoryTags([e.currentTarget.value as CategoryTags])
```
@poofytoo It is unclear to me based on the mocks if you intended for the tag selection to act as a radio button and only allow one selection at a time or a set of checkboxes where multiple tags could be selected.

This current behavior looks to be radio buttons. If that is the desired behavior, no change necessary.

dillzhang replied:
If this is the intentional behavior, we should `searchCategoryTags` to be the category we are showing rather than a list.

poofytoo replied:
The intention was originally to be checkboxes - but after looking and playing around with this, I'm ok with how this is :) thank you!

dillzhang commented on pages/electronics-library-dev.tsx:
```diff
@@ -109,6 +86,59 @@ const Home = ({
+    if (searchCategoryTags[0] !== CategoryTags.ShowAll) {
+      componentList = componentList.filter((component) => {
+        return component.category?component.category.toLowerCase().includes(searchCategoryTags[0].toLowerCase()):null;
+      });
```
Given that the `component`'s category is a string, why are we checking `includes` rather than `===`?

mmtywa replied:
I left as includes jic we did end up going with checkboxes/multiple selections. I'm good with either, just let me know which you'd prefer.

mmtywa replied:
Resolution:
![image](https://user-images.githubusercontent.com/129784422/231253249-e652b02d-2f92-4852-b80a-91af00d79fc6.png)

dillzhang commented on pages/electronics-library-dev.tsx:
```diff
@@ -109,6 +86,59 @@ const Home = ({
+    }
+
+    if(searchField) {      
+      const searcher = new Searcher(componentList,{keySelector: (obj:IElectronicsComponent) => [obj.category, obj.name, obj.shortDescription, obj.description, obj.tags?.join(" "), obj.shownTo, obj.leadTime]});
```
A precursory glance at the docs for `fast-fuzzy` seems to suggest that results will be returned based on their best match, which is not the described behavior on the Figma where name should be prioritized before shortDescription which should be prioritized before description, etc. You can see this by searching `32`, which has `Temperature & Humidity` as the first result instead of `Adafruit HUZZAH32 (ESP32)`.

@poofytoo Not sure how much you care about that behavior.

poofytoo replied:
this is ok! I would leave this.

dillzhang commented on pages/electronics-library-dev.tsx:
```diff
@@ -109,6 +86,59 @@ const Home = ({
+        <span style={{color: categoryObject[searchCategoryTags[0]].color, fontWeight: 700}}>
+          {categoryObject[searchCategoryTags[0]].result}
+        </span>
+        <span style={{transition:"0.2s all ease", opacity: searchField?'1':'0'}}>
+          {' '}with "{searchField}"
+        </span>
```
I don't think animation is necessary. We can just conditionally render it instead since the text is already jumping when `Showing` and `Searching all` change. 

Something like `{searchField && <span>...</span>}`

mmtywa replied:
Good call. I originally had that animation in before the debouncer since the text updating while typing was distracting. I put in the fix you recommended. Thanks!

dillzhang commented on pages/electronics-library-dev.tsx:
```diff
@@ -109,6 +86,59 @@ const Home = ({
+
+  const [timerId, setTimerId] = useState<ReturnType<typeof setTimeout>>();
+  useEffect(() => {
+    filterComponents();
```
We should move this into the `setTimeout` call as well. If we are going to debounce the updates to the `ResultsHeader`, we should definitely debounce the searching through components. This will help with performance as well when we have a longer list of components.

dillzhang replied:
I realize with my comment above about changing `ResultsHeader` to a functional component, debouncing here no longer makes sense.

The proper place to debounce would be `onSearchChange` function. If we debounce there, we can simply add `searchField` to the list of dependencies in the `useEffect` above.

mmtywa replied:
Alright got this implemented:
- Timer/debouncer got pulled into onSearchChange.
- Removed the useEffect on searchField.
- Made buildResultsHeader a functional component and pulled out of page scope.

This definitely cleaned things up. Solid recommendations. Learning some better practices with this one. Thanks! :+1:

dillzhang commented on pages/electronics-library-dev.tsx:
```diff
@@ -109,6 +86,59 @@ const Home = ({
+  }
+
+  const [resultsHeader, setResultsHeader] = useState<JSX.Element>();
+  const buildResultsHeader = () => {
```
Love that you pulled this into a helper function. 

I believe the proper pattern for this would be creating a functional component for `ResultsHeader` and passing in the necessary arguments (`searchField`) as props. This way you do not have to use a state to store this JSX.Element.

mmtywa replied:
fixed with revision above

dillzhang commented on pages/electronics-library-dev.tsx:
```diff
@@ -128,8 +158,39 @@ const Home = ({
+
+      <div className={styles.searchBar}>
+        <label htmlFor="search">Search:</label>
+        <input aria-label='Search Electronic Components' type='search' name='search' onChange={onSearchChange} />
```
add `id="search"` so click on the label will focus on the input as well

mmtywa replied:
Good catch

dillzhang commented on pages/electronics-library-dev.tsx:
```diff
@@ -150,6 +211,19 @@ const Home = ({
         ))}
       </div>
+
+      { filteredComponents.length ? null :
```
`filteredComponents.length === 0 && <p .... `

dillzhang commented on styles/Page.module.scss:
```diff
@@ -72,6 +153,15 @@
 }
 
+.askTas {
+  width: 690px;
```
@poofytoo did you want this specific width or are we fine with it running to natural page width?

poofytoo replied:
this should be natural page width

mmtywa replied:
Wasn't sure if this was a text readability thing in the mock. Removed :+1:

dillzhang commented on pages/electronics-library-dev.tsx:
```diff
@@ -128,8 +158,39 @@ const Home = ({
+
+      <div className={styles.resultsHeader}>
+        <h2>{resultsHeader}</h2>
+        <h3>{filteredComponents.length===1?`${filteredComponents.length} result`:`${filteredComponents.length} results`}</h3>
```
Might be easier to read if we are simply appending the `s` to `result`

`result{filterComponents.length===1 ? '' : 's'}`

dillzhang reviewed (commented):
Looks pretty good to me, @poofytoo has final say over UI.

Left some comments and about behavior and coding patterns.

poofytoo commented on pages/electronics-library-dev.tsx:
```diff
@@ -18,7 +21,23 @@ interface IElectronicsComponent {
+};
+
+enum CategoryTags {
+  ShowAll = "ShowAll",
```
Add space between Show and All

mmtywa replied:
enum didn't like when I added spaces. Here's the fix I put in. The Tag should have a space now. Good catch!
![image](https://user-images.githubusercontent.com/129784422/231242337-3c3dd1f3-ed33-480d-ab24-58e37baf6d61.png)

poofytoo commented on pages/electronics-library-dev.tsx:
```diff
@@ -128,8 +158,39 @@ const Home = ({
       </div>
+
+      <div className={styles.searchBar}>
+        <label htmlFor="search">Search:</label>
```
Add roughly 3px space below label (same with *show:*)

poofytoo commented on styles/Page.module.scss:
```diff
@@ -9,8 +9,89 @@
+  height: 34px;
+  background: #FFFFFF;
+  border: 2px solid #464646;
+  border-radius: 0.5rem;
```
change to 0.4rem;

poofytoo commented on styles/Page.module.scss:
```diff
@@ -9,8 +9,89 @@
+  color: #5A5A5A;
+}
+
+.searchBar input{
```
I think input by default does not inherit font family - can you add `font-family: "DM Sans", sans-serif` and `font-style: 1rem`

mmtywa replied:
Good catch! :)

poofytoo commented on pages/electronics-library-dev.tsx:
```diff
@@ -128,8 +158,39 @@ const Home = ({
+
+      <div className={styles.searchBar}>
+        <label htmlFor="search">Search:</label>
+        <input aria-label='Search Electronic Components' type='search' name='search' onChange={onSearchChange} />
```
Oh wow. I don't think I've ever realized you could do type='search'. That's awesome.

mmtywa replied:
I didn't know that either, but I saw it in a resource I was referencing for this search bar :)

poofytoo reviewed (changes requested):
Some changes, please! Also, out of curiosity, is it best practice to do a `p` for `2p00b` because it's a period (supposed to be 2.00b)? or did you just unintentionally type `p` haha

mmtywa commented:
> Some changes, please! Also, out of curiosity, is it best practice to do a `p` for `2p00b` because it's a period (supposed to be 2.00b)? or did you just unintentionally type `p` haha

Yeah this was just habit of mine. I've banged my head one too many times with issues with periods in titles/file names, that I just take them out. I can put it back in :+1:

mmtywa committed:
Adds revisions from review
- rgb colors to hex
- linting
- moved debounce into onSearchChange
- buildResultsHeader made functional component and pulled out of page scope
- removed fixed width on askTas
- css cleanup

dillzhang reviewed (approved):
LGTM. Please make the change from `include` to `===` that we discussed on slack and merge when ready.

poofytoo reviewed (approved):
LGTM!

mmtywa committed:
Cleans up filter logic by CategoryTag
- swapped includes which had incorrect logic for strict equal
